R/bs4_book.R

```diff
@@ -515,6 +516,54 @@ template_link <- function(html, xpath, href) {
   }
 }
 
+tweak_metadata <- function(html, path) {
+  file <- basename(path)
+  if (file == "index.html") {
+    return(invisible())
+  } else {
+    # Fix og:url
+    og_url <- xml2::xml_find_first(html, '//meta[@property="og:url"]')
+    base_url <- xml2::xml_attr(og_url, "content")
+    if (!grepl("/$", base_url)) {
+      base_url <- paste0(base_url, "/")
+    }
+    xml2::xml_set_attr(og_url, "content", paste0(base_url, file))
+    # Fix description
+    og_description <- xml2::xml_find_first(html, '//meta[@property="og:description"]')
+
+    if (file == "table-of-contents.html") {
+      xml2::xml_set_attr(
+        og_description,
+        "content",
+        paste("Table of contents;", xml2::xml_attr(og_description, "content"))
+      )
+    } else {
+      contents <- copy_html(xml2::xml_find_first(html, "//main[@id='content']"))
+      xml2::xml_remove(xml2::xml_find_first(contents, "//h1"))
+      xml2::xml_remove(xml2::xml_find_first(contents, "//div[@class='chapter-nav']"))
+      text <- xml2::xml_text(contents)
+      text <- gsub("\\\n", " ", text)
+      text <- gsub("  ", " ", text)
+      text <- gsub("^[[:space:]]+", "", text)
+      text <- gsub("[[:space:]]+$", "", text)
+      if (nzchar(text)) {
+        words <- unlist(strsplit(text, " "))
+        no_char <- cumsum(unlist(lapply(words, function(x) {nchar(x) + 1})))
+        max_n <- max(which(no_char<= 197))
+        desc <- paste(words[1: max_n], collapse = " ")
+        if (max_n != length(words)) {
+          desc <- paste0(desc, "...")
+        }
+        xml2::xml_set_attr(og_description, "content", desc)
```

Comment:
I've now tweaked the code.

When one creates a book with the RStudio menu, there's a default description.

Reply:
> When one creates a book with the RStudio menu, there's a default description.

So you are assuming here that every book will be created this way and a description will already be set in the YAML so that the HTML node are created in headers ? 

What seems to be the wy it works now: 
* If no `description` is set in YAML header already, then none will be set in the output. 
* If a `description` is set in YAML, then the content won't matter because it will be overwritten by an auto generated description based on the content. 

Am I right ? 

It does not seem like a good API to set  a text in description so that it is ignored 🤔

Reply:
I'll change this now.